File: Binset.java
package com.mycompany.myapp;

public class Binset
{
	private final boolean isSet = true;
	public Binset one, two;
	
	public Binset() { one = null; two = null; }
	public Binset(int i){
		if(i <= 0)
		{ one = null; two = null; }
		else if(i == 0)
		{ one = new Binset();
		  two = new Binset();
		}
		else
		{ one = new Binset((i + 1)/2);
		  two = new Binset(i/2);			
		}
	}
	public Binset(Binset b)
	{
		if(b.isEmpty())
		{ one = null; two = null; }
		else if(b.isOne())
		{ one = new Binset();
		  two = new Binset(); }
		else
		{ one = new Binset(b.one);
		  two = new Binset(b.two); }
	}
	public Binset(Binset a, Binset b){
		if(b.isEmpty())
		{
			if(a.isEmpty())
			{one = null; two = null; }
			else if(a.isOne())
			{ one = new Binset();
			  two = new Binset(); }
			else
			{ one = new Binset(a.one);
			  two = new Binset(a.two); }
		}
		else{
			one = new Binset(a);
			two = new Binset(b);
		}
	}
	public boolean isEmpty()
	{ return (one == null)
	  && (two == null); }
	public boolean isOne(){
		return !isEmpty() &&
		one.isEmpty() && two.isEmpty();
	}
	public int Count(){
		if(isEmpty()) return 0;
		if(isOne()) return 1;
		return one.Count() +
		two.Count();
	}
	public boolean setEquals(Binset b){
		if(this == b) return true;
		if(isEmpty()) return b.isEmpty();
		if(b.isEmpty()) return false;
		return one.setEquals(b.one) &&
		two.setEquals(b.two);
	}
	public Binset andWith(Binset b){
		if(isEmpty() || b.isEmpty())
			return new Binset();
		if(isOne())
		{
			if(b.isOne())
				return new Binset(1);
			return andWith(b.one);
		}
		if(b.isOne())
			return one.andWith(b);
		return new Binset(
			one.andWith(b.one),
			two.andWith(b.two)
		);
	}
	public boolean Contains(Binset b)
	{
		return andWith(b).setEquals(b);
	}
	public Binset excWith(Binset b){
		if(isEmpty() || b.isEmpty())
			return new Binset(this);
		if(b.Contains(this))
			return new Binset();
		if(isOne())
			return new Binset(1);
		if(b.isOne())
			return new Binset(
				one.excWith(b),
				two
			);
		return new Binset(
			one.excWith(b.one),
			two.excWith(b.two)
		);
	}
	public Binset xorWith(Binset b)
	{
		if(isEmpty())
			return new Binset(b);
		if(b.isEmpty())
			return new Binset(this);
		if(isOne())
		{
			if(b.isOne())
				return new Binset();
			return new Binset(xorWith(b.one), b.two);
		}
		if(b.isOne())
			return new Binset(one.xorWith(b), two);
		return new Binset(
			one.xorWith(b.one),
			two.xorWith(b.two)
		);
	}
	public Binset orWith(Binset b){
		if(isEmpty())
			return new Binset(b);
		if(b.isEmpty())
			return new Binset(this);
		if(isOne())
		{
			if(b.isOne())
				return new Binset(1);
			return new Binset(
				orWith(b.one),
				b.two
			);
		}
		if(b.isOne())
	 		return new Binset(
				one.orWith(b),
				two
			);
		return new Binset(
			one.orWith(b.one),
			two.orWith(b.two)
		);
	}
	public Binset plus(Binset b){
		if(isEmpty())
			return new Binset(b);
		if(b.isEmpty())
			return new Binset(this);
		if(isOne())
		{
			if(b.isOne())
				return new Binset(2);
			return new Binset(
				plus(b.one),
				b.two
			);
		}
		if(b.isOne())
			return new Binset(
				one.plus(b),
				two
			);
		return new Binset(
			one.plus(b.one),
			two.plus(b.two)
		);
	}
	public Binset times(Binset b){
		if(isEmpty() || b.isEmpty())
			return new Binset();
		if(isOne())
			return new Binset(b);
		if(b.isOne())
			return new Binset(this);
		return new Binset(
			one.times(b),
			two.times(b)
		);
	}
}

File: binset.py
class binset:
    def __init__(self, in1=None, in2=None):
        if(in1 is None):
            self.one=None
            self.two=None
        elif(isinstance(in1,binset)):
            if(isinstance(in2,binset)):
                self.one=binset(in1)
                self.two=binset(in2)
            elif(in1.isEmpty()):
                self.one=None
                self.two=None
            elif(in1.isOne()):
                self.one=binset()
                self.two=binset()
            else:
                self.one=binset(in1.one)
                self.two=binset(in1.two)
        elif(isinstance(in1,int)):
            if(in1==0):
                self.one=None
                self.two=None
            elif(in1==1):
                self.one=binset()
                self.two=binset()
            else:
                self.one=binset(in1//2 + in1%2)
                self.two=binset(in1//2)
    def isEmpty(self):
        return (self.one is None) and (self.two is None)
    def isOne(self):
        return not self.isEmpty() and self.one.isEmpty() and self.two.isEmpty()
    def count(self):
        if(self.isEmpty()):
            return 0
        elif(self.isOne()):
            return 1
        else:
            return self.one.count() + self.two.count()
    def __eq__(self,other):
        if(self is other):
            return true
        if(type(self)!=type(other)):
            return False
        if(self.isEmpty()):
            return other.isEmpty()
        if(other.isEmpty()):
            return False
        if(self.isOne()):
            return other.isOne()
        if(other.isOne()):
            return False
        return (self.one==other.one) and \
               (self.two==other.two)
    def __and__(self,other):
        if not isinstance(other,binset):
            return None
        if(self.isEmpty() or other.isEmpty()):
            return binset()
        if(self.isOne()):
            if(other.isOne()):
                return binset(1)
            return self & other.one
        if(other.isOne()):
            return self.one & other
        return binset(self.one & other.one, \
                      self.two & other.two)
    def contains(self,other):
        return (self & other)==other
    def __xor__(self,other):
        if not isinstance(other,binset):
            return self
        if(self.isEmpty()):
           return binset(other)
        if(other.isEmpty()):
            return binset(self)
        if(self.isOne()):
            if(other.isOne()):
                return binset()
            return binset(self ^ other.one,other.two)
        if(other.isOne()):
            return binset(self.one ^ other,self.two)
        return binset(self.one ^ other.one, \
                      self.two ^ other.two)
    def __or__(self,other):
        if not isinstance(other,binset):
            return self
        if(self.isEmpty()):
            return binset(other)
        if(other.isEmpty()):
            return binset(self)
        if(self.isOne()):
            if(other.isOne()):
                return binset(1)
            return binset(self | other.one,other.two)
        if(other.isOne()):
            return binset(self.one | other,self.two)
        return binset(self.one | other.one, self.two | other.two)
    def __sub__(self,other):
        if not isinstance(other,binset):
            return None
        if(self.isEmpty() or other.contains(self)):
            return binset()
        if(other.isEmpty()):
            return binset(self)
        if(self.isOne()):
            return self - other.one
        if(other.isOne()):
            return binset(self.one-other,self.two)
        return binset(self.one - other.one, \
                      self.two - other.two)
    def __add__(self,other):
        if not isinstance(other,binset):
            return self
        if(self.isEmpty()):
            return binset(other)
        if(other.isEmpty()):
            return binset(self)
        if(self.isOne()):
            if(other.isOne()):
                return binset(2)
            return binset(self + other.one,other.two)
        if(other.isOne()):
            return binset(self.one + other,self.two)
        return binset(self.one + other.one, \
                      self.two + other.two)
    def __mul__(self,other):
        if not isinstance(other,binset):
            return self
        if(self.isEmpty() or other.isEmpty()):
            return binset()
        if(self.isOne()):
            return binset(other)
        if(other.isOne()):
            return binset(self)
        return binset(self.one * other, \
                      self.two * other)
    def __str__(self):
        if(self.isEmpty()):
            return '0'
        elif(self.isOne()):
            return '1'
        else:
            return self.one.__str__()+':'+self.two.__str__()

class binarray(binset):
    def __init__(self,size:int=2,T:type=binset):
        super().__init__()
        self.size=size if isinstance(size,int) and (size>2) else 2
        self.T=T if issubclass(T,binset) else binset
    def __getitem__(self,ix:int):
        if((ix<0) or (ix>=self.size)):
            raise IndexError()
        if(self.size==2):
            if(ix==0):
                return self.one if self.one is not None else binset()
            return self.two if self.two is not None else binset()
        if((self.size==3) and (ix==2)):
            return self.two if self.two is not None else binset()
        if(ix<(self.size+1)//2):
            if isinstance(self.one,binarray):
            	return self.one[ix]
            return binset()
        if isinstance(self.two,binarray):
            return self.two[ix-(self.size + 1)//2]
        return binset()
    def __setitem__(self,ix:int,value):
        if((ix<0) or (ix>=self.size)):
            raise IndexError()
        if(self.size==2):
        	if ix==0:
        		if (value is None) and isinstance(self.two,self.T):
        			self.one=binset()
        			return
        		self.one=value
        		if value and not isinstance(self.two,binset):
        			self.two=binset()
        	else:
        		if (value is None) and isinstance(self.one,self.T):
        			self.two=binset()
        			return
        		self.two=value
       			if value and not isinstance(self.one,binset):
        			self.one=binset()
        	return
        if((self.size==3) and (ix==2)):
            if (value is None) and isinstance(self.one,binarray):
            	self.two=binset()
            	return
            self.two=value
            if value and not isinstance(self.one,binset):
                self.one=binset()
            return
        if(ix<(self.size + 1)//2):
            if not isinstance(self.one,binarray):
            	if not value:
            		return
            	self.one=binarray((self.size+1)//2,self.T)
            if isinstance(value,self.T):
            	if not isinstance(self.two,binset):
                	self.two=binset()
            elif not isinstance(self.two,self.T):
            	self.one=None
            	self.two=None
            	return
            self.one[ix]=value
        else:
            if not isinstance(self.two,binarray):
            	if not value:
            		return
            	self.two=binarray(self.size//2,self.T)
            if isinstance(value,self.T):
            	if not isinstance(self.one,binset):
                    self.one=binset()
            elif not isinstance(self.one,self.T):
            	self.one=None
            	self.two=None
            	return
            self.two[ix-(self.size+1)//2]=value
    def len(self):
        if(self.isOne()):
            return 0
        ct=0
        if(isinstance(self.one,binarray)):
            ct+=self.one.len()
        elif(isinstance(self.one,T)):
            ct+=1
        if(isinstance(self.two,binarray)):
            ct+=self.two.len()
        elif(isinstance(self.two,T)):
            ct+=1
        return ct
    def __iter__(self):
    	self.iterix=0;
    	return self
    def __next__(self):
    	ret=None
    	while not ret:
    		if self.iterix>=self.size:
    			raise StopIteration
    		ret=self[self.iterix]
    		if not isinstance(ret,self.T):
     			ret = None
    		self.iterix+=1
    	return ret
